add-on/chrome/bootstrap.js

```diff
@@ -968,6 +1070,75 @@ var WindowListener = {
 };
 
 /**
+ * Provide a way to throttle functionality using DNS to distribute 3 numbers for
+ * various distributions channels. DNS is used to scale distribution of the
+ * numbers as an A record pointing to a loopback address (127.*.*.*). Prefs are
+ * used to control behavior (what domain to check) and keep state (a ticket
+ * number to track if it needs to initialize, to wait for its turn, or is
+ * completed).
+ */
+let Throttler = {
+  // Each 8-bit block of the IP address allows for 0% rollout (value 0) to 100%
+  // rollout (value 255).
+  TICKET_LIMIT: 255,
+
+  // Allow the DNS service to be overwritten for testing.
+  _dns: Cc["@mozilla.org/network/dns-service;1"].getService(Ci.nsIDNSService),
+
+  /**
+   * Check if a given feature should be throttled or not.
+   * @param {string} [prefPrefix] Start of the preference name for the feature.
+   * @return {Promise} Resolved on success, and rejected on throttled.
+   */
+  check(prefPrefix) {
+    return new Promise((resolve, reject) => {
+      // Initialize the ticket (0-254) if it doesn't have a valid value yet.
+      let prefTicket = prefPrefix + ".ticket";
+      let ticket = Services.prefs.getIntPref(prefTicket);
+      if (ticket < 0) {
+        ticket = Math.floor(Math.random() * this.TICKET_LIMIT);
+        Services.prefs.setIntPref(prefTicket, ticket);
+      }
+      // Short circuit if the special ticket value indicates we're good to go.
+      else if (ticket >= this.TICKET_LIMIT) {
+        resolve();
+        return;
+      }
+
+      // Handle responses from the DNS resolution service request.
+      let onDNS = (request, record) => {
+        // Use a specific part of the A-record IP address depending on the
+        // channel. I.e., 127.[release/beta/other].[aurora].[nightly].
+        let index = 1;
+        switch (Services.prefs.getCharPref("app.update.channel")) {
+          case "aurora":
+            index = 2;
+            break;
+          case "nightly":
+            index = 3;
+            break;
+        }
+
+        // Check if the 8-bit threshold (0-255) exceeds the ticket (0-254).
+        let threshold = record && record.getNextAddrAsString().split(".")[index];
```

Comment:
While exceedingly elegant, this line has more than one layer of abstractions, and depends on knowing that IPv4 addresses when passed to split this way will create the array you want.  Please at least split this up with temporaries for readability.